portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -43,36 +46,62 @@
    */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int maxComments = Integer.parseInt(UtilityFunctions.getFieldFromResponse(request, "maxcomments", defaultMaxComment));
     String sortMetric = UtilityFunctions.getFieldFromResponse(request, "metric", "time");
     String sortOrder = UtilityFunctions.getFieldFromResponse(request, "order", "desc");
 
+    ArrayList<UserComment> comments = new ArrayList<>();
+    comments = populateComments(comments, 0, maxComments, sortOrder, sortMetric);
+
+    Gson gson = new Gson();
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(comments));
+  }
+
+  /*
+   * Returns a list containing atmost maxComments top-level queries and all their replies. 
+   * The queries are sorted by sortMetric in sortOrder
+   */ 
+  private ArrayList<UserComment> populateComments(ArrayList<UserComment> comments, long rootId, int maxComments, String sortOrder, String sortMetric) {
     Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
     Builder builder = Query.newEntityQueryBuilder();
-    builder = builder.setKind("Comment");
+    builder = builder.setKind("Comment").setFilter(PropertyFilter.eq("rootid", rootId));
+
     if (sortOrder.equals("desc")) {
       builder = builder.setOrderBy(StructuredQuery.OrderBy.desc(sortMetric));
     } else {
       builder = builder.setOrderBy(StructuredQuery.OrderBy.asc(sortMetric));
     }
+
+    if (rootId == 0) {
+      builder = builder.setLimit(maxComments);
+    }
+
     Query<Entity> query = builder.build();
     QueryResults<Entity> results = datastore.run(query);
-
-    ArrayList<UserComment> comments = new ArrayList<>();
     while (results.hasNext()) {
       Entity entity = results.next();
-      long id = entity.getKey().getId();
-      String name = entity.getString("name");
-      String email = entity.getString("email");
-      long time =  entity.getLong("time");
-      String comment = entity.getString("comment");
-      long parentId = entity.getLong("parentid");
-      UserComment userComment = UserComment.create(name, email, comment, time, id, parentId);
-      comments.add(userComment);
+      UserComment comment = entityToComment(entity);
+      comments.add(comment);
+      if (rootId == 0) {
+        long newRootId = entity.getKey().getId();
+        comments = populateComments(comments, newRootId, maxComments, sortOrder, sortMetric);
```

Comment:
Instead of having a recursive function that always only recurse once, I would suggest creating two separate functions, populateRootComments & populateChildrenComments. This way you don't have to have a bunch of `if (rootId == 0)` checks.